Blog post for the student project in Jenkins security (#4419)
jenkins-infra/jenkins.io: content/blog/2021/06/2021-06-17-student-luminy-project-security.adoc

---
layout: post
title: "Four students and their master project in Jenkins security"
tags:
- plugins
- security
- jenkins
authors: 
- jphilip
- kguerroudj
- qparra
- mheyries
opengraph:

>   image: /images/post-images/2020/08-external-fingerprint-storage-phase-3/opengraph.png

Reviewer: We need a different opengraph image.
```suggestion
  image: /images/post-images/2020/08-external-fingerprint-storage-phase-3/opengraph.png
```

That image describes an [external fingerprint storage](https://www.jenkins.io/images/post-images/2020/08-external-fingerprint-storage-phase-3/opengraph.png) project.

Attached is a possible idea for an opengraph image (see the [original source](https://docs.google.com/presentation/d/1Q1PgNnRTgzBpVRXPqQo3PudzCa2eoc6_1_NRjFRMLrU/edit#slide=id.ge0af3de53e_0_0) if you'd like to modify it).

![Jenkins Social Media Covers (1)](https://user-images.githubusercontent.com/156685/122239090-e5688c00-ce7d-11eb-8722-c959ea41013b.png).

---

=== Context

> Jenkins is the world’s leading open-source automation server used by companies large and small around the globe to implement continuous integration and continuous delivery.

Reviewer 2: kind of redundant in the Jenkins blog

=== Project Goals

The main goal of this project was to allow us to discover and work on diverse security vulnerabilities in the context of a widely-used software solution, and in order to achieve that we had to separate the project in multiple other goals:

* Learning about some of the most common vulnerabilities and the form in which they can be found on a Jenkins Instance and its plugins.

* Performing code review and technical audit on the application, and compiling our results as part of a security team.

* Reporting our findings to the Jenkins team and the plugin maintainers, while sometimes helping the latter to correct these vulnerabilities.

=== Knowledge Sharing

At the beginning of the project, we set up communication channels with our mentor through Slack and Google Meets, and agreed to schedule weekly sessions with him.
The goal of these sessions was both to teach us more about the functionalities of Jenkins and the types of vulnerabilities we would encounter, and to allow us to ask more specific questions regarding our findings.

> Thanks to our mentor developing a mock plugin compiling a variety of classic vulnerabilities and several of their implementations, including SSRFs, XSSs, and XXEs, we have been able to learn through practice.

Reviewer 2: > SSRFs, XSSs, and XXEs

Probably means nothing to most readers, perhaps a Wikipedia link?

> Thanks to our mentor developing a mock plugin compiling a variety of classic vulnerabilities and several of their implementations, including SSRFs, XSSs, and XXEs, we have been able to learn through practice.

Reviewer: I agree with Daniel.  I had to lookup the descriptions of those acronyms.  I would prefer if the acronyms were expanded on first reference.  Possible approach:
```suggestion
Thanks to our mentor developing a mock plugin compiling a variety of classic vulnerabilities and several of their implementations, including server side request forgery ([SSRF](https://en.wikipedia.org/wiki/Server-side_request_forgery)), cross site scripting ([XSS](https://owasp.org/www-community/attacks/xss/)), and XML external entity ([XXE](https://en.wikipedia.org/wiki/XML_external_entity_attack)) attacks, we have been able to learn through practice.
```

It allowed us to analyze the context of the code and the different ways the Jelly framework can be used to display information, expanding our payload options and giving us a clearer view of the patterns to look for during our code reviews.

We have also had the opportunity to learn about the process used to report the vulnerabilities to the maintainers through Jira issues, and some ways we could correct them or provide steps to do so.

=== Searching and compiling

> At first, we decided to work as a pretty loose team, with each member working on a different plugin and regrouping our findings to confirm or infirm them, while staying in constant contact to ask each other questions.

Reviewer: I've not seen the word "[infirm](https://www.merriam-webster.com/dictionary/infirm)" used to describe the rejection of a hypothesis.  Possibly means "reject"?
```suggestion
At first, we decided to work as a pretty loose team, with each member working on a different plugin and regrouping our findings to confirm or reject them, while staying in constant contact to ask each other questions.
```

During the last third of this project, we began to work together on bigger plugins, in order to have more points of view reflecting on the same problem.
With different analyses, we were able to come up with new payloads, and sometimes with new vulnerabilities where we only found one separately.

=== Reporting and correcting the vulnerabilities

All of the reporting was done through Jira issues, which allows the Jenkins team to centralize and triage the vulnerabilities.
Once we provided the necessary information, along with the reproduction steps we had, a member of the Jenkins security team contacted the plugin maintainer and guided them through the next steps of the process, with hope that they would answer.

We have also tried to make the maintainers job easier, working on some fixes.
To achieve this, we delved not only into the functionalities of the vulnerable plugins, but also into some mitigation processes that we found either in the Jenkins documentation,  or with the help of our mentor.

> Each one of our modifications have been tested locally, in order to assess whether the vulnerability was still present, and whether no function had been altered.

Reviewer 2: ```suggestion
Each one of our modifications has been tested locally, in order to assess whether the vulnerability was still present, and whether no function had been altered.
```
